lib/src/main/kotlin/tsl/trees/RBTree.kt

```diff
@@ -0,0 +1,322 @@
+                uncle = currentParent?.parent?.rightChild
+                if (uncle?.color == RBNode.Color.Red) {
+                    currentParent?.color = RBNode.Color.Black
+                    uncle.color = RBNode.Color.Black
+                    currentGrandParent?.color = RBNode.Color.Red
+                    newNode = currentGrandParent
+                } else {
+                    if (
+                        currentParent?.color == RBNode.Color.Red &&
+                            newNode.color == RBNode.Color.Red &&
+                            newNode == currentParent.rightChild
+                    ) {
+                        newRoot = rotateLeft(currentParent, newRoot)
+                        newNode.color = RBNode.Color.Black
+                        newNode.leftChild?.color = RBNode.Color.Red
+                        newRoot = rotateRight(currentGrandParent, newRoot)
+                        newNode.rightChild?.color = RBNode.Color.Red
+                        break
+                    }
+                    currentParent?.color = RBNode.Color.Black
+                    currentGrandParent?.color = RBNode.Color.Red
+                    newRoot = rotateRight(currentGrandParent, newRoot)
+                }
+            } else if (currentParent == currentGrandParent?.rightChild) {
+                uncle = newNode.parent?.parent?.leftChild
+                if (uncle?.color == RBNode.Color.Red) {
+                    currentParent?.color = RBNode.Color.Black
+                    uncle.color = RBNode.Color.Black
+                    currentGrandParent?.color = RBNode.Color.Red
+                    newNode = currentGrandParent
+                } else {
+                    if (
+                        currentParent?.color == RBNode.Color.Red &&
+                            newNode.color == RBNode.Color.Red &&
+                            newNode == currentParent.leftChild
+                    ) {
+                        currentParent.leftChild?.color = RBNode.Color.Black
+                        newRoot = rotateRight(currentParent, newRoot)
+                    }
+                    currentParent?.color = RBNode.Color.Black
+                    currentGrandParent?.color = RBNode.Color.Red
+                    newRoot = rotateLeft(currentGrandParent, newRoot)
+                }
+            }
+        }
+        newRoot?.color = RBNode.Color.Black
+        root = newRoot
+    }
+
+    /**
+     * Returns something and sets up the root depending on the result of deleteNode.\.
+     *
+     * @param key The key of the node to be deleted\.
+     * @return Value if delete was successful and null if it isn't.\.
+     */
+    public override fun delete(key: K): V? {
+        val deletingNodeValue = search(key)
+        root = deleteNode(key)
+        if (root != null) return deletingNodeValue
+        root?.color = RBNode.Color.Black // ensure the root is black after deletion
```

Comment:
Метод delete будет переработан подобным образом. 
`public override fun delete(key: K): V? {
        val deletingNodeValue = search(key)
        if (deletingNodeValue == null) return null // in case we do not found the deletinnode
        root = deleteNode(key)
        root?.color = RBNode.Color.Black // ensure the root is black after deletion
        return deletingNodeValue
    }`